File: mobile-android/app/src/main/java/ph/attendance/sync/Protocol.kt
package ph.attendance.sync

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import ph.attendance.domain.AttendanceRecord
import ph.attendance.domain.AttendanceStatus
import ph.attendance.domain.Gender
import ph.attendance.domain.Schedule
import ph.attendance.domain.SchoolDay
import ph.attendance.domain.Section
import ph.attendance.domain.Student
import ph.attendance.domain.TimeWindow

/**
 * The wire contract, matching `backend/src/protocol.ts` field for field.
 *
 * Deliberately separate from the domain types. The local model is free to change whenever the app
 * wants; the wire format cannot, because a phone that has not been updated in a month still has
 * to sync with a server the web client also talks to.
 */

const val PROTOCOL_VERSION = 1

@Serializable
data class WireSection(
    val id: String,
    val name: String,
    val archived: Boolean,
    val createdAt: String,
    val updatedAt: String,
) {
    fun toDomain() = Section(id, name, archived, createdAt, updatedAt)

    companion object {
        fun from(s: Section) = WireSection(s.id, s.name, s.archived, s.createdAt, s.updatedAt)
    }
}

@Serializable
data class WireStudent(
    val id: String,
    val sectionId: String,
    val studentNumber: String,
    val lastName: String,
    val firstName: String,
    val middleName: String = "",
    val gender: String,
    val archived: Boolean,
    val createdAt: String,
    val updatedAt: String,
) {
    fun toDomain() = Student(
        id, sectionId, studentNumber, lastName, firstName, middleName,
        Gender.fromWire(gender), archived, createdAt, updatedAt,
    )

    companion object {
        fun from(s: Student) = WireStudent(
            s.id, s.sectionId, s.studentNumber, s.lastName, s.firstName, s.middleName,
            s.gender.wire, s.archived, s.createdAt, s.updatedAt,
        )
    }
}

@Serializable
data class WireWindow(val start: String, val end: String)

@Serializable
data class WireSchedule(
    val id: String,
    val sectionId: String,
    val title: String,
    val venue: String = "",
    val present: WireWindow,
    val late: WireWindow,
    val archived: Boolean,
    val createdAt: String,
    val updatedAt: String,
) {
    fun toDomain() = Schedule(
        id, sectionId, title, venue,
        TimeWindow(present.start, present.end), TimeWindow(late.start, late.end),
        archived, createdAt, updatedAt,
    )

    companion object {
        fun from(s: Schedule) = WireSchedule(
            s.id, s.sectionId, s.title, s.venue,
            WireWindow(s.present.start, s.present.end), WireWindow(s.late.start, s.late.end),
            s.archived, s.createdAt, s.updatedAt,
        )
    }
}

@Serializable
data class WireRecord(
    val id: String,
    val scheduleId: String,
    val sectionId: String,
    val studentNumber: String,
    val date: String,
    val status: String,
    val scheduleTitle: String = "",
    val recordedAt: String,
) {
    fun toDomain() = AttendanceRecord(
        id, scheduleId, sectionId, studentNumber, date,
        AttendanceStatus.fromWire(status), scheduleTitle, recordedAt,
    )

    companion object {
        fun from(r: AttendanceRecord) = WireRecord(
            r.id, r.scheduleId, r.sectionId, r.studentNumber, r.date,
            r.status.wire, r.scheduleTitle, r.recordedAt,
        )
    }
}

@Serializable
data class WireSchoolDay(val date: String, val firstSeenAt: String) {
    fun toDomain() = SchoolDay(date, firstSeenAt)

    companion object {
        fun from(d: SchoolDay) = WireSchoolDay(d.date, d.firstSeenAt)
    }
}

@Serializable
data class ChangeSet(
    val sections: List<WireSection> = emptyList(),
    val students: List<WireStudent> = emptyList(),
    val schedules: List<WireSchedule> = emptyList(),
    val records: List<WireRecord> = emptyList(),
    val schoolDays: List<WireSchoolDay> = emptyList(),
) {
    val size: Int
        get() = sections.size + students.size + schedules.size + records.size + schoolDays.size
}

@Serializable
data class CreateWorkspaceRequest(val name: String)

@Serializable
data class JoinWorkspaceRequest(val joinCode: String)

@Serializable
data class WorkspaceResponse(
    val workspaceId: String,
    val name: String,
    val joinCode: String? = null,
    val token: String,
)

@Serializable
data class PullRequest(val since: Long = 0, val limit: Int = 500)

@Serializable
data class PullResponse(
    val changes: ChangeSet,
    val cursor: Long,
    val hasMore: Boolean,
    val serverTime: String,
)

@Serializable
data class PushRequest(val changes: ChangeSet)

@Serializable
data class PushResponse(
    val applied: Int,
    val skipped: Int,
    val cursor: Long,
    val serverTime: String,
)

@Serializable
data class SyncStatusResponse(
    val configured: Boolean,
    @SerialName("protocol") val protocolVersion: Int = PROTOCOL_VERSION,
)

@Serializable
data class ApiError(val error: String, val code: String)

File: backend/src/protocol.ts
import { z } from "zod";

/**
 * The wire contract between a device and the sync server.
 *
 * Defined here rather than imported from the web app's `domain/` on purpose. The local model is
 * free to change shape whenever the app wants; the wire format cannot, because a phone that has
 * not been updated in a month still has to sync. Keeping them separate makes that independence
 * explicit instead of accidental.
 */

export const PROTOCOL_VERSION = 1 as const;

/* ------------------------------------------------------------------- shared */

const uuid = z.uuid();
const timestamp = z.iso.datetime();
const isoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Date must be YYYY-MM-DD");
const time24 = z.string().regex(/^([01]\d|2[0-3]):[0-5]\d$/, "Time must be HH:mm");
const studentNumber = z
  .string()
  .min(1)
  .max(32)
  .regex(/^[A-Za-z0-9][A-Za-z0-9._-]*$/);
const shortText = (max: number) => z.string().max(max);

/* ------------------------------------------------------------------ entities */

export const wireSectionSchema = z.object({
  id: uuid,
  name: shortText(80).min(1),
  archived: z.boolean(),
  createdAt: timestamp,
  updatedAt: timestamp,
});

export const wireStudentSchema = z.object({
  id: uuid,
  sectionId: uuid,
  studentNumber,
  lastName: shortText(60).min(1),
  firstName: shortText(60).min(1),
  middleName: shortText(60),
  gender: z.enum(["male", "female"]),
  archived: z.boolean(),
  createdAt: timestamp,
  updatedAt: timestamp,
});

export const wireScheduleSchema = z.object({
  id: uuid,
  sectionId: uuid,
  title: shortText(80).min(1),
  venue: shortText(80),
  present: z.object({ start: time24, end: time24 }),
  late: z.object({ start: time24, end: time24 }),
  archived: z.boolean(),
  createdAt: timestamp,
  updatedAt: timestamp,
});

export const wireRecordSchema = z.object({
  id: uuid,
  scheduleId: uuid,
  sectionId: uuid,
  studentNumber,
  date: isoDate,
  status: z.enum(["present", "late", "absent"]),
  scheduleTitle: shortText(80),
  recordedAt: timestamp,
});

export const wireSchoolDaySchema = z.object({
  date: isoDate,
  firstSeenAt: timestamp,
});

export type WireSection = z.infer<typeof wireSectionSchema>;
export type WireStudent = z.infer<typeof wireStudentSchema>;
export type WireSchedule = z.infer<typeof wireScheduleSchema>;
export type WireRecord = z.infer<typeof wireRecordSchema>;
export type WireSchoolDay = z.infer<typeof wireSchoolDaySchema>;

/* ------------------------------------------------------------------- batches */

/**
 * A push is capped well below anything a real school produces in one batch. Without a ceiling,
 * a single request could ask the server to validate and upsert an unbounded number of rows.
 */
const MAX_PER_PUSH = 2_000;

export const changeSetSchema = z.object({
  sections: z.array(wireSectionSchema).max(MAX_PER_PUSH).default([]),
  students: z.array(wireStudentSchema).max(MAX_PER_PUSH).default([]),
  schedules: z.array(wireScheduleSchema).max(MAX_PER_PUSH).default([]),
  records: z.array(wireRecordSchema).max(MAX_PER_PUSH).default([]),
  schoolDays: z.array(wireSchoolDaySchema).max(MAX_PER_PUSH).default([]),
});

export type ChangeSet = z.infer<typeof changeSetSchema>;

export function emptyChangeSet(): ChangeSet {
  return { sections: [], students: [], schedules: [], records: [], schoolDays: [] };
}

export function changeSetSize(changes: ChangeSet): number {
  return (
    changes.sections.length +
    changes.students.length +
    changes.schedules.length +
    changes.records.length +
    changes.schoolDays.length
  );
}

/* ------------------------------------------------------------------ requests */

export const createWorkspaceRequestSchema = z.object({
  name: shortText(80).min(1),
});

export const joinWorkspaceRequestSchema = z.object({
  joinCode: z.string().min(8).max(32),
});

export const workspaceCredentialsSchema = z.object({
  workspaceId: uuid,
  name: z.string(),
  /** Shown once, so another device can join. Never returned again. */
  joinCode: z.string().optional(),
  token: z.string(),
});

export type WorkspaceCredentials = z.infer<typeof workspaceCredentialsSchema>;

/**
 * The cursor is a server-assigned sequence number, not a timestamp.
 *
 * Device clocks disagree — sometimes by minutes, occasionally by years. A timestamp cursor
 * silently skips rows written while the clock was wrong. A sequence assigned by the database is
 * monotonic by construction, so "everything after N" means exactly that.
 */
export const pullRequestSchema = z.object({
  since: z.number().int().min(0).default(0),
  limit: z.number().int().min(1).max(1000).default(500),
});

export const pullResponseSchema = z.object({
  changes: changeSetSchema,
  cursor: z.number().int().min(0),
  hasMore: z.boolean(),
  serverTime: timestamp,
});

export const pushRequestSchema = z.object({
  changes: changeSetSchema,
});

export const pushResponseSchema = z.object({
  applied: z.number().int().min(0),
  skipped: z.number().int().min(0),
  cursor: z.number().int().min(0),
  serverTime: timestamp,
});

export type PullRequest = z.infer<typeof pullRequestSchema>;
export type PullResponse = z.infer<typeof pullResponseSchema>;
export type PushRequest = z.infer<typeof pushRequestSchema>;
export type PushResponse = z.infer<typeof pushResponseSchema>;

/* -------------------------------------------------------------------- errors */

export const errorResponseSchema = z.object({
  error: z.string(),
  code: z.enum([
    "unauthorized",
    "not_found",
    "invalid_request",
    "rate_limited",
    "unavailable",
  ]),
});

export type ErrorResponse = z.infer<typeof errorResponseSchema>;

/**
 * Last-write-wins, with a deterministic tiebreak.
 *
 * Two devices editing the same row while both offline is normal, not exceptional. Whoever wrote
 * last wins; if the timestamps are identical — clocks are coarse, and a bulk edit can produce
 * the same millisecond — the larger id wins. Arbitrary, but identical on every device and on the
 * server, which is the only property that matters. Without it, replicas can disagree forever.
 */
export function incomingWins(
  incoming: { updatedAt: string; id: string },
  existing: { updatedAt: string; id: string },
): boolean {
  if (incoming.updatedAt !== existing.updatedAt) {
    return incoming.updatedAt > existing.updatedAt;
  }
  return incoming.id > existing.id;
}
